### PR TITLE
fix generate openAPI failed.

### LINKED ISSUE #283: OpenApi generation for registered services fail when application started from jar
When the application is started via spring boot fat jar, then the open API generation is failing.

Currently the lib is not copied into the spring boot jar.

17:08:48.683 [thread-pool-1] ERROR o.e.x.orchestrator.utils.OpenApiUtil []- Error: Unable to access jarfile ./lib/openapi-generator-cli.jar
17:08:48.689 [thread-pool-1] ERROR o.e.x.orchestrator.utils.OpenApiUtil []- Create service openApi html file failed.Error: Unable to access jarfile ./lib/openapi-generator-cli.jar

### PR DESCRIPTION
**Pull request by baixinsui**
fix #283

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,6 +6,6 @@ ENV TERRAFORM_VERSION=1.4.4
 RUN wget https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_linux_amd64.zip
 RUN unzip terraform_${TERRAFORM_VERSION}_linux_amd64.zip
 RUN mv terraform /usr/bin/terraform
-
+COPY runtime/target/lib/ lib/
```
Comment (swaroopar) [inline Dockerfile:9]: I think its better we download the file here also directly from internet and delete the jar from our codebase.

Comment (baixinsui) [reply Dockerfile]: Ok. I deleted the jar from codebase and download it into the workdir by URL.

File: runtime/src/main/resources/application.properties [hunk 2]
```diff
@@ -18,6 +18,7 @@ spring.h2.console.settings.web-allow-others=true
 spring.jackson.deserialization.fail-on-unknown-properties=true
 spring.hateoas.use-hal-as-default-json-media-type=false
 terraform.workspace.directory=xpanse_deploy_ws
+openapi.download-generator-client-url=https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/6.5.0/openapi-generator-cli-6.5.0.jar
```
Comment (iskey) [inline runtime/src/main/resources/application.properties:21]: Just some suggestion:
can we add a property named `openapi.generator-client-version`?
If the `openapi.download-generator-client-url` specified, this package will be used.
If the  `openapi.generator-client-version` specified, we should get the specified version.
else we should get the latest version.

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 3]
```diff
@@ -211,41 +250,73 @@ public String createServiceApi(RegisterServiceEntity registerService) {
...
-     * @param registerService Registered services.
-     */
-    @Async("taskExecutor")
-    public void generateServiceApi(RegisterServiceEntity registerService) {
-        createServiceApi(registerService);
     }
 
     /**
-     * delete OpenApi for registered service using the ID.
+     * Download openapi-generator-cli.jar from remoteURL into workdir.
```
Comment (iskey) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:280]:
```suggestion
     * Download openapi-generator-cli.jar from the maven repository or the URL specified by the `@openapi.download-generator-client-url` into the work directory.
```

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 4]
```diff
@@ -211,41 +250,73 @@ public String createServiceApi(RegisterServiceEntity registerService) {
         }
     }
 
+
     /**
-     * update OpenApi for registered service .
+     * Get open API workdir.
      *
-     * @param registerService Registered services.
+     * @return openApi workdir
```
Comment (iskey) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:257]:
```suggestion
     * @return the work directory of the openApi.
```

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 5]
```diff
@@ -211,41 +250,73 @@ public String createServiceApi(RegisterServiceEntity registerService) {
         }
     }
 
+
     /**
-     * update OpenApi for registered service .
+     * Get open API workdir.
```
Comment (iskey) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:255]:
```suggestion
     * Get the work directory of the openAPI.
```

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 6]
```diff
@@ -136,13 +147,40 @@ public String getOpenApiUrl(String id) {
...
+     *
+     * @param registerService Registered services.
+     */
+    @Async("taskExecutor")
+    public void generateServiceApi(RegisterServiceEntity registerService) {
+        createServiceApi(registerService);
+    }
+
+    /**
+     * delete OpenApi for registered service using the ID.
```
Comment (iskey) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:174]:
```suggestion
     * Delete OpenApi document for registered service using the ID.
```

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 7]
```diff
@@ -136,13 +147,40 @@ public String getOpenApiUrl(String id) {
...
+    public void updateServiceApi(RegisterServiceEntity registerService) {
+        File file = new File(getOpenApiWorkdir(), registerService.getId() + ".html");
+        if (file.exists()) {
+            log.info("Delete old openApi file:{}, success:{}", file.getName(), file.delete());
+        }
+        createServiceApi(registerService);
+    }
+
+    /**
+     * update OpenApi for registered service .
```
Comment (iskey) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:164]:
```suggestion
     * Generate/update OpenApi document for registered service .
```

File: runtime/src/main/resources/application.properties (hunk 2 above)
Comment (baixinsui) [reply runtime/src/main/resources/application.properties:21]: There is no method to get lastest version of the generator-client-cli.jar from  the maven repo. I can use openapi.generator-client-version replace the openapi.download-generator-client-url in the properties file and join the complete URL with version in the OpenApiUtils.class. Is it Ok？

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java [hunk 8]
```diff
@@ -211,41 +250,73 @@ public String createServiceApi(RegisterServiceEntity registerService) {
...
-    public void generateServiceApi(RegisterServiceEntity registerService) {
-        createServiceApi(registerService);
     }
 
     /**
-     * delete OpenApi for registered service using the ID.
+     * Download openapi-generator-cli.jar from remoteURL into workdir.
      *
-     * @param id ID of registered service.
+     * @param workdir workdir
```
[the remark this answers is not in the document]
Comment (baixinsui) [reply modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/utils/OpenApiUtil.java:282]: OK.

**Review by iskey — APPROVED** [review]
looks good.

**Pull request merged**